sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/ReturnInFinallyCheck.kt

```diff
@@ -0,0 +1,130 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtBreakExpression
+import org.jetbrains.kotlin.psi.KtContinueExpression
+import org.jetbrains.kotlin.psi.KtDoWhileExpression
+import org.jetbrains.kotlin.psi.KtElement
+import org.jetbrains.kotlin.psi.KtExpressionWithLabel
+import org.jetbrains.kotlin.psi.KtFinallySection
+import org.jetbrains.kotlin.psi.KtForExpression
+import org.jetbrains.kotlin.psi.KtLabeledExpression
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.KtReturnExpression
+import org.jetbrains.kotlin.psi.KtThrowExpression
+import org.jetbrains.kotlin.psi.KtTreeVisitorVoid
+import org.jetbrains.kotlin.psi.KtWhileExpression
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+/**
+ * Checks if the label in the element points to an unexpected label
+ */
+private fun isEscapingTheBlock(element: KtExpressionWithLabel, expectedLabels: List<String>): Boolean {
+    val label = element.getLabelName() ?: return false
+    return label !in expectedLabels
+}
+
+private fun buildReportMessage(element: KtElement) =
+    "Remove this %s statement from this finally block.".format(element.toString().lowercase())
+
+@Rule(key = "S1143")
+class ReturnInFinallyCheck : AbstractCheck() {
+
+    override fun visitFinallySection(finallySection: KtFinallySection, kotlinFileContext: KotlinFileContext) {
+        val report: (KtElement) -> Unit = {
+            kotlinFileContext.reportIssue(it, buildReportMessage(it))
+        }
+
+        val visitor = FinallyBlockVisitor(report)
+        finallySection.accept(visitor)
+        visitor.jumpsToLabel
+            .filter { isEscapingTheBlock(it, visitor.labelsInFinallyBlock) }
+            .forEach(report)
+
+        finallySection
+            .collectDescendantsOfType<KtThrowExpression>()
+            .forEach(report)
+    }
+
+    private class FinallyBlockVisitor(private val report: (KtElement) -> Unit) : KtTreeVisitorVoid() {
+        private var loopDepthCounter = 0
+        private var lambdaDepthCounter = 0
+        val labelsInFinallyBlock = ArrayList<String>()
+        val jumpsToLabel = ArrayList<KtExpressionWithLabel>()
+
+        override fun visitLabeledExpression(expression: KtLabeledExpression) {
+            expression.name?.let { labelsInFinallyBlock.add(it) }
+            super.visitLabeledExpression(expression)
+        }
+
+        override fun visitLambdaExpression(lambdaExpression: KtLambdaExpression) {
+            lambdaDepthCounter++
+            super.visitLambdaExpression(lambdaExpression)
+            lambdaDepthCounter--
```

Comment:
I think it will be nice to extract a repeated patterns into a separate function, to make code more concise